[AMBARI-24364] - Mpack Instance Manager Not Listing Versions Correctly

scottduan opened the pull request:
## What changes were proposed in this pull request?
After merge, mpack_manager, execution_command and some other python module codes are broken

## How was this patch tested?

1. UT test
2. Manually e2e test

scottduan committed:
AMBARI-24364: Mpack Instance Manager Not Listing Versions Correctly -- Part 1

scottduan committed:
AMBARI-24364: Mpack Instance Manager Not Listing Versions Correctly

scottduan commented on ambari-common/src/main/python/resource_management/libraries/execution_command/execution_command.py:
```diff
@@ -96,7 +96,8 @@ def get_component_instance_name(self):
     Retrieve service name from command.json, eg. 'zk1'
     :return: service name
     """
-    if '_CLIENTS' in self.get_module_name(): # FIXME temporary hack
```
module_name is not always available

scottduan commented on ambari-common/src/main/python/resource_management/libraries/execution_command/execution_command.py:
```diff
@@ -279,14 +280,20 @@ def get_group_list(self):
     :return: a list of users
     """
-    return self.__get_value("stackSettings/user_list")
+    user_list = self.__get_value("stackSettings/user_list")
```
Now the command.json does not include stack_name, stack_version, user_list, group_list etc in stackSettings, i.e if I restart installed zookeeper service, the command.json has no such fields. We may discuss these fields and the locations later. At this time, just follow the old way.

scottduan commented on ambari-common/src/main/python/resource_management/libraries/functions/get_not_managed_resources.py:
```diff
@@ -34,7 +34,9 @@ def get_not_managed_resources():
   """
   config = Script.get_config()
-  not_managed_hdfs_path_list = json.loads(config['stackSettings']['not_managed_hdfs_path_list'])[:]
+  not_managed_hdfs_path_list = []
```
Same, stackSettings may not have this field.

scottduan commented on ambari-common/src/main/python/resource_management/libraries/functions/mpack_manager_helper.py:
```diff
@@ -31,19 +31,16 @@
+  return get_conf_dir(mpack_name, instance_name, subgroup_name, module_name,
                            {components_instance_type: [component_instance_name]})
 
-  return conf_json[COMPONENTS_PLURAL_KEY_NAME][components_instance_type.lower()][COMPONENT_INSTANCES_PLURAL_KEY_NAME][
```
This code has totally broken, there are two issues here:
1. No way to access conf_json dict directly via "components" key, you have to use "mpacks/mpack_name/...", but they are not always passed in, these arguments can be None
2. The original code return a string only for single mpack instance, but we actually need consider mutiple mpack instances, so the new code return a list

scottduan commented on ambari-common/src/main/python/resource_management/libraries/script/script.py:
```diff
@@ -216,10 +216,11 @@ def save_component_version_to_structured_out(self):
 
     try:
       mpack_version, component_version = mpack_manager_helper.get_versions(mpack_name,
-        mpack_instance_name, module_name, component_type )
```
This is to fix a traceback issue

scottduan commented on mpack-instance-manager/src/main/python/instance_manager/instance_manager.py:
```diff
@@ -108,9 +109,18 @@ def get_conf_dir(mpack=None, mpack_instance=None, subgroup_name=DEFAULT_SUBGROUP
-  return build_granular_json_with_filtering(mpack, mpack_instance, subgroup_name, module_name, components_map,
-                                            output_conf_dir=True)
+  dirs, is_client = get_conf_log_run_dir_helper(mpack, mpack_instance, subgroup_name, module_name, components_map, output_conf_dir = True, output_log_dir = False, output_run_dir = False)
+  if not is_client:
```
Please see my inline comments. The layout of client component is different server component, so we have to handle them separately

scottduan commented on mpack-instance-manager/src/main/python/instance_manager/instance_manager.py:
```diff
@@ -149,12 +165,70 @@ def get_run_dir(mpack=None, mpack_instance=None, subgroup_name=DEFAULT_SUBGROUP_
+
+
+def walk_mpack_dict(mpack_dict, filter, dirs):
+  """
```
In multiple mpack instances scenario, we need traverse the whole mpack dict and get all the paths that we want.

scottduan commented:
The corresponding zookeeper code review:

https://github.com/hortonworks/zookeeper-PDR/pull/7

@jonathan-hurley @adoroszlai @ncole @swapanshridhar @jayush  Please review the code and let me know your thoughts. Thanks.

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/Ambari-Github-PullRequest-Builder/3375/
Test FAILed.
Test FAILured.

jonathan-hurley commented on ambari-common/src/main/python/resource_management/libraries/functions/mpack_manager_helper.py:
```diff
@@ -65,24 +60,22 @@ def get_component_rundir_path(mpack_name, instance_name, module_name, components
+          (mpack_name == None or mpack_name.lower() in dir) and (instance_name == None or instance_name.lower() in dir)]
 
 
 def get_versions(mpack_name, instance_name, module_name, components_instance_type,
```
I think you also need to make changes to this method, no?

scottduan replied:
I think so, I will also add a new UT there. Thanks.

jonathan-hurley reviewed (approved):
Can you please verify if you need to do anything with the get_version method?

scottduan committed:
AMBARI-24364: Mpack Instance Manager Not Listing Versions Correctly

scottduan commented:
@jonathan-hurley @swapanshridhar I have updated the code to include get_version() in mpack_manager_helper.py, also added a test case in test_instance_manager.py.

scottduan commented:
@jonathan-hurley @swapanshridhar Please also do a quick review for https://github.com/hortonworks/zookeeper-PDR/pull/7,

Just two line of code review. I need to merge both PRs at the same time. Thanks.

jonathan-hurley commented:
@scottduan my concern right now is that there's a lot of work being put on the python code. For example:

```
instance_name=mpack_instance_name, module_name=module_name,
zk_log_dir = ""
if zk_log_dirs:
  zk_log_dir = zk_log_dirs[0]
zk_pid_dirs = get_component_rundir_path...
```

Can we do this in the helper so that we don't have to pollute all of the Python code?

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/Ambari-Github-PullRequest-Builder/3397/
Test FAILed.
Test FAILured.

scottduan committed:
AMBARI-24364: Mpack Instance Manager Not Listing Versions Correctly

scottduan commented:
@jonathan-hurley I have updated the code and at this time I assume there is only one single mpack instance, so the mpack_manager_helper only return single path for config_dir, log_dir and run_dir.

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/Ambari-Github-PullRequest-Builder/3437/
Test FAILed.
Test FAILured.

scottduan commented:
Checked the db:

ambari=> select * from hostcomponentstate;
 id | host_comp_desired_state_id | cluster_id |  component_name  |  component_type  | mpack_version |   version   | current_state | last_live_state | host_id | service_group_id | service_id | upgrade_st
ate 
----+----------------------------+------------+------------------+------------------+---------------+-------------+---------------+-----------------+---------+------------------+------------+-----------
----
  2 |                          2 |          2 | ZOOKEEPER_CLIENT | ZOOKEEPER_CLIENT | 1.0.0-b607    | 3.4.0.0-b42 | INSTALLED     | INSTALLED       |       1 |                2 |          2 | NONE
  1 |                          1 |          2 | ZOOKEEPER_SERVER | ZOOKEEPER_SERVER | 1.0.0-b607    | 3.4.0.0-b42 | STARTED       | STARTED         |       1 |                2 |          3 | NONE
(2 rows)

ambari=> 
 
All versions fields are populated.
